### ISSUE TITLE
Setting project title or layer name in Russian

### ISSUE THREAD

**Issue by Kluichnikov** [tomchadwin/qgis2web#199]
Hi. I have recently opened for myself Qgis2web. It's a great thing for a common mapper. I like it very much. The only thing I'd like to see in Qgis2web is the ability to use project title or layer name written in Russian. Now I ain't able to use these things in localized version.

For example if I use project title written in Russian I see the [following message (in Leaflet case)](http://pastebin.com/9E5B0d86).

If I use layer's name with the Russian characters I see [another message (in Leaflet case)](http://pastebin.com/9zP5PqMY).

Is there some roundabout in order to solve these two issues of mine?
Kind regards, Eugene Kluichnikov (Russian Federation, Moscow)

**Comment by tomchadwin** [comment]
I've committed a couple of changes to master. Can you download and test? They might have fixed your immediate issues, but more internationalization errors are probably still present.

**Comment by michellemoine** [comment]
Have the same issue with French characters in layers name. It's fine in OL3.

**Comment by michellemoine** [comment]
Just avoided special French characters to keep going.

**Comment by tomchadwin** [comment]
@michellemoine Error still present with current master? Can you paste the error, so I can find which line numbers cause the issue?

**Comment by Kluichnikov** [comment]
Thank you so much for the fast reply. The only problem for me that I dunno how to install the master version of qgis2web. I tried to find that out by myself but still wasn't able to find the way for installing the master.

**Comment by michellemoine** [comment]
Here you are:
UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 14732: ordinal not in range(128)

**Comment by tomchadwin** [comment]
To install master:

- browse to https://github.com/tomchadwin/qgis2web
- click "Download ZIP", and open the zip file
- copy the contents of `qgis2web-master` to the qgis2web folder in your QGIS plugins folder
- restart QGIS

**Comment by michellemoine** [comment]
Of course, it only happens if I choose to display the layers list.

**Comment by tomchadwin** [comment]
@michellemoine Sorry, but I need the whole thing: the line number I need is in one of the error lines above that.

**Comment by michellemoine** [comment]
Error

qgis2web produced an error:

Traceback (most recent call last):
File "C:/.qgis2/python/plugins\qgis2web\maindialog.py", line 125, in previewMap
MainDialog.previewLeaflet(self)
File "C:/.qgis2/python/plugins\qgis2web\maindialog.py", line 308, in previewLeaflet
visible, cluster, json, params, popup)
File "C:/.qgis2/python/plugins\qgis2web\leafletWriter.py", line 186, in writeLeaflet
matchCRS, canvas, mapLibLocation, new_src, template)
File "C:/.qgis2/python/plugins\qgis2web\leafletFileScripts.py", line 151, in writeHTMLstart
f_html.write(base)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 14732: ordinal not in range(128)

**Comment by tomchadwin** [comment]
Try current master - line 151 now reads:

```
f_html.write(unicode(base))```

**Comment by michellemoine** [comment]
I'll download it again

**Comment by tomchadwin** [comment]
Apologies - my mistake. I hadn't committed. It's up there now.

**Comment by michellemoine** [comment]
Hers is a caption on the file I have:
with open(outputIndex, 'w') as f_html:
        base = replaceInTemplate(template + ".html", values)
        base = re.sub('\n[\s_]+\n', '\n', base)
        f_html.write(unicode(base))
        f_html.close()

The same error is still showing

**Comment by Kluichnikov** [comment]
Hi Tom. I see another error message in a separate window when I used the last master version. I used just the project title written in Russian. [Here is a link to an error message](http://pastebin.com/Gc0Uu2ed)

**Comment by michellemoine** [comment]
Added this to LeafletFileScripts.py

import sys
reload(sys)
sys.setdefaultencoding('utf8')

Seems to work.

**Comment by tomchadwin** [comment]
Hrm:

http://stackoverflow.com/questions/3828723/why-should-we-not-use-sys-setdefaultencodingutf-8-in-a-py-script

Might have to find another solution.

**Comment by tomchadwin** [comment]
@michellemoine, can you remove your code, and instead add the following to the beginning of `leafletScriptStrings.py`:

```
# -*- coding: latin-1 -*-
```

Must be the first line of the script. If you still have the error, add it to `leafletLayerScripts.py` as well. Can you let us know if it helps?

**Comment by Kluichnikov** [comment]
Tom, I used the recipe by Michelle helped me. I mean the lines added to LeafletFileScripts.py:

> import sys
> reload(sys)
> sys.setdefaultencoding('utf8')

**Comment by tomchadwin** [comment]
@Kluichnikov See the link above - this is not recommended. Can you try my suggestion above, instead?

**Comment by Kluichnikov** [comment]
Alright, Tom. I will try your way now.

**Comment by Kluichnikov** [comment]
I have tried, Tom, you way. And I see the next message:

> Error
> 
> qgis2web produced an error:
> 
> Traceback (most recent call last):
> File "C:/Users/kluchnikov.e/.qgis2/python/plugins\qgis2web\maindialog.py", line 125, in previewMap
> MainDialog.previewLeaflet(self)
> File "C:/Users/kluchnikov.e/.qgis2/python/plugins\qgis2web\maindialog.py", line 308, in previewLeaflet
> visible, cluster, json, params, popup)
> File "C:/Users/kluchnikov.e/.qgis2/python/plugins\qgis2web\leafletWriter.py", line 169, in writeLeaflet
> new_src += unicode(titleStart)
> UnicodeDecodeError: 'ascii' codec can't decode byte 0xd0 in position 278: ordinal not in range(128)

**Comment by Kluichnikov** [comment]
I added "# -*- coding: latin-1 -*-" to "leafletScriptStrings.py" file at first. An then to "leafletLayerScripts.py" file also.

**Comment by tomchadwin** [comment]
I'm not sure how best to solve this. It's possible that adding a second argument to each use of `unicode()` might help:

```
unicode(text, "utf-8")
```

If you are able to try that, I'd be grateful.

**Comment by Kluichnikov** [comment]
Where ought I to search for the use of unicode(), Tom? In "leafletScriptStrings.py" and "leafletLayerScripts.py" files only. Or I ought to search this text in the other py-files also?

**Comment by tomchadwin** [comment]
It's used in lots of Python files, but I think changing lots of them or all of them at once is a bad idea. I'd try changing the ones listed in the error messages might be worth trying.

**Comment by Kluichnikov** [comment]
Okay, gotcha. I will only try to make changes in the files mentioned in the error messages. Thank you one more time so much for your help.

**Comment by tomchadwin** [comment]
Thanks for the offer of helping to solve this one. I do want to warn you that I feel that the problem is much more deep-seated than this, and that this approach will probably not solve it. I just don't know enough about Unicode yet to make definite recommendations.

@akbargumbira, I don't know if you have any observations which could help us on this issue?

**Comment by akbargumbira** [comment]
yes, changing system's encoding is not recommended. I will take a look at this

**Comment by akbargumbira** [comment]
@tomchadwin this was my strategy to support unicode in InaSAFE. You might want to have a look

**Comment by Kluichnikov** [comment]
@tomchadwin Tom, I tried changing unicode() statements in different files using different combinations. Still I didn't manage to make qgis2web work in case if my file had got the Russian characters inside. So I am ready to give up by now. I do hope you will be able to resolve this issue in the future.

**Comment by michellemoine** [comment]
Hi Tom,
I've tried your suggestions without success.

**Comment by akbargumbira** [comment]
I just made a PR here https://github.com/tomchadwin/qgis2web/pull/201. It should fix the problem when exporting layers with accented chars to Leaflet

**Comment by tomchadwin** [comment]
Many thanks, @akbargumbira! @Kluichnikov, can you get your Russian project and layer names to work now?

**Comment by michellemoine** [comment]
Thank you guys, it's working fine for French.

**Comment by tomchadwin** [comment]
Splendid. I just need to check there are no other instances of writing text to disk which need to be changed in the same way. For instance, how about layer attribute titles and values? Do they work with non-ASCII characters? It's probably all covered by @akbargumbira's commit, but let me know if you spot anything else.

**Comment by tomchadwin** [comment]
I just realized that my suggestion - the comment at the beginning of the script - specified the wrong encoding: Latin-1 when it should be UTF-8. Never mind - apologies, everyone.

**Comment by Kluichnikov** [comment]
@tomchadwin @akbargumbira Good morning, guys. I tried all the ways you proposed here and in the PR #201 but I still ain't able to use qgis2web while using the project title or the layers names written in Russian. I see [the following message.](http://pastebin.com/VNnYx7sB) If I click the "Update preview" button anyway I get [another message](http://pastebin.com/n7PJxas4) in a separate window.

What did I try to do?

- I used the pure master version;
- Added "leafletFileScripts.py" from PR #201;
- Added "# -*- coding: utf-8 -*-" to "leafletScriptStrings.py" and "leafletLayerScripts.py".
- ---------------------------------------------------------------------------------
- I installed the base version of qgis2web;
- Added the master vesrion to the base version;
- Added leafletFileScripts.py from PR #201;
- Added "# -*- coding: utf-8 -*-" to "leafletScriptStrings.py" and "leafletLayerScripts.py

**Comment by akbargumbira** [comment]
Hi @Kluichnikov, can you just master branch of this repo and tell me the detail of your environments:
- If possible, provide the data to replicate the issue
- What settings in QGIS2web you are using

**Comment by Kluichnikov** [comment]
Hi @akbargumbira. I used your last recomendation and used just the master branch repo. [Here is a link to my data and images](https://drive.google.com/folderview?id=0B1RNSNTkf_wEY056UlQwUzhSZzg&usp=sharing) depicting my settings in qgis2web. The same images with my settings can be seen below also. Kindest regards.

English title & two English layers
![eng_title_layers](https://cloud.githubusercontent.com/assets/3087476/12843688/09bf23c2-cc0c-11e5-996f-b8413d8e0165.jpg)

English title & an English layer & a Russian layer
![rus_one_layer](https://cloud.githubusercontent.com/assets/3087476/12843708/335fcfba-cc0c-11e5-90fb-ba0983a630e2.jpg)

English title & two Russian layers
![rus_two_layers](https://cloud.githubusercontent.com/assets/3087476/12843721/40ab47d0-cc0c-11e5-96eb-23f967bbb464.jpg)

Russian project title & two English layers
![rus_title](https://cloud.githubusercontent.com/assets/3087476/12843738/5e24e92e-cc0c-11e5-8d04-bb036cf8d37b.jpg)

**Comment by tomchadwin** [comment]
Can you try removing the `unicode()` function from l.169 of `leafletWriter.py`, making it:

```
        new_src += titleStart
```

**Comment by tomchadwin** [comment]
@michellemoine Do you get the same error if you set your project title including French characters?

**Comment by Kluichnikov** [comment]
I tried removing the unicode() function from the line 169. The result has remained just the same.
![errorwindow](https://cloud.githubusercontent.com/assets/3087476/12845075/78bd8e00-cc14-11e5-838b-84f6a82d0cd4.jpg)

Please, take note of the case when I had two layers named in Russian. Then there was just one layer depicted in the qgis2web window instead of two ones.
![errorwindow2](https://cloud.githubusercontent.com/assets/3087476/12845146/fbfce004-cc14-11e5-8ef8-e1c13cf59e09.jpg)

**Comment by tomchadwin** [comment]
I've recreated the title error, and am looking into that now.

I've also looked at the two Russian layers problem, and that is caused by the function safeName() in utils.py. As you can see from the comment (which was written before qgis2web existed, in qgis-ol3), the javascript variables are derived from the layer names, but only the ASCII characters are retained. Because the Russian layer names have no ASCII characters in them, the var names have zero length. Although they are then prefixed and suffixed throughout the code, your two layers end up with identical var names. This obviously does not work. I'm not sure how to solve this yet, but will have a think.

**Comment by tomchadwin** [comment]
See the issue originally raised for qgis2leaf: https://github.com/Geolicious/qgis2leaf/issues/202. Perhaps using `layer.id()` for JS vars is an option...

**Comment by Kluichnikov** [comment]
Thank you very much, Tom. I will have a look at that issue now.

**Comment by tomchadwin** [comment]
I've committed a fix to master which perhaps solves the two Russian layers issue. However, it destroys all the tests (output has timestamped vas, so is never equal to the control files. I probably therefore need to reconsider how to do this.

**Comment by tomchadwin** [comment]
OK, I've switched to an approach which I think I can make work with the tests. Please download current master, and see if you can now have two layers with Russian names in your layers list.

Not solved the project title issue yet. Will fix the tests first, and then look at that.

**Comment by michellemoine** [comment]
Hi all,

I am still using the master version of yesterday and i have the same error if the title includes French characters. However, it's working nicely in the layers legend titles.

Error

qgis2web produced an error:

Traceback (most recent call last):
File "C:/.qgis2/python/plugins\qgis2web\maindialog.py", line 125, in previewMap
MainDialog.previewLeaflet(self)
File "C:/.qgis2/python/plugins\qgis2web\maindialog.py", line 308, in previewLeaflet
visible, cluster, json, params, popup)
File "C:/.qgis2/python/plugins\qgis2web\leafletWriter.py", line 169, in writeLeaflet
new_src += unicode(titleStart)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 318: ordinal not in range(128)

**Comment by tomchadwin** [comment]
OK, I've committed a fix for Unicode project titles to master. Can you test?

**Comment by Kluichnikov** [comment]
Please, excuse me. I was engaged for a long time due to my work. After I've downloaded the last master I see that everything works. The project title and the layers written in Russian are seen in the ready web-page.
![untitled-1](https://cloud.githubusercontent.com/assets/3087476/12850532/3892c1fc-cc36-11e5-9b9f-ee52d5da6cb5.jpg)

**Comment by tomchadwin** [comment]
Great. Glad it's fixed. @michellemoine, let me know if it works for you as well. Closing for now.

**Comment by tomchadwin** [comment]
Released as 1.0.0-RC3 and pushed to QGIS plugins repo

**Comment by michellemoine** [comment]
It's working fine. Thanks Tom.
